packages/styletron-inferno/src/styled.js

```diff
@@ -0,0 +1,124 @@
+const createElement = require('inferno-create-element');
+const {injectStylePrefixed} = require('styletron-utils');
+const {
+  assign,
+  isNil,
+  isObject,
+  isString,
+  isFunction
+} = require('./utils');
+
+const STYLETRON_KEY = '__STYLETRON';
+
+module.exports = styled;
+
+/**
+ * Helper function to create styled components
+ * @packagename styletron-inferno
+ * @param  {string|function} name   Tag name or component function/class
+ * @param  {function|object} styles Style object or function that returns a style object
+ * @return {function}               Styled component
+ *
+ * @example
+ * import { styled } from 'styletron-inferno';
+ *
+ * const Panel = styled('div', {
+ *   backgroundColor: 'lightblue',
+ *   fontSize: '12px'
+ * });
+ *
+ * <Panel>Hello World</Panel>
+ *
+ * @example
+ * import { styled } from 'styletron-inferno';
+ *
+ * const Panel = styled('div', (props) => ({
+ *   backgroundColor: props.alert ? 'orange' : 'lightblue',
+ *   fontSize: '12px'
+ * }));
+ *
+ * <Panel alert>Danger!</Panel>
+ *
+ * @example
+ * import { styled } from 'styletron-inferno';
+ *
+ * const DeluxePanel = styled(Panel, (props) => ({
+ *   backgroundColor: props.alert ? 'red' : 'lime',
+ *   boxShadow: '3px 3px 3px darkgray',
+ *   color: 'white'
+ * }));
+ *
+ * <DeluxePanel>Bonjour Monde</DeluxePanel>
+ */
+function styled(name, styles) {
+
+  // Styled component
+  if (name && name.hasOwnProperty(STYLETRON_KEY)) {
+    const component = name[STYLETRON_KEY];
+    const stylesArray = component.stylesArray.concat(styles);
+    return createStyledComponent(component.name, stylesArray);
+
+  // Tag name or non-styled component
+  } else if (isString(name) || isFunction(name)) {
+    return createStyledComponent(name, [styles]);
+  }
+
+  throw new Error('Must pass a tag name or component');
+}
+
+function createStyledComponent(name, stylesArray) {
+
+  function StyledComponent(props, context) {
+    const newProps = assign({}, props);
+    const styles = resolveStyles(stylesArray, props, context);
+    const className = injectStylePrefixed(context.styletron, styles);
+    newProps.className = mergeClassNames(props.className, className);
+    if (isFunction(props.innerRef)) {
+      newProps.ref = props.innerRef;
+      delete newProps.innerRef;
+    }
+    return createElement(name, newProps);
+  }
+
+  StyledComponent[STYLETRON_KEY] = {name, stylesArray};
+
+  return StyledComponent;
+}
+
+function resolveStyles(stylesArray, props, context) {
+  const resolvedStyles = {};
+  for (let i = 0, l = stylesArray.length, styles; i < l; i++) {
+    styles = stylesArray[i];
+    if (!isNil(styles)) {
+      if (isFunction(styles)) {
+        assign(resolvedStyles, styles(props, context));
+      } else if (isObject(styles)) {
+        assign(resolvedStyles, styles);
+      }
+    }
+  }
+  return resolvedStyles;
+}
+
+const SPACE_REGEX = /\s+/;
+
+function mergeClassNames(classNameA, classNameB) {
```

Comment:
I confess....the only reason I added this function was for test purposes 😁 . When composing styled components (`styled(AlreadyStyledComponent)`) I was seeing `className="a a b c"` which was inconsistent with React (I'm guessing React does class de-duplication?).

So I guess the question then becomes—"does it matter?". 

Do you know if there is an overhead in the browser when parsing an element's class list? Does having the same class repeated mean that the browser will parse and apply that style twice?

If it doesn't, then we can probably just drop it—I agree that it has a somewhat unnecessary cost to it. However, if it does have a browser parsing cost, then let's optimise it as you suggested above.